Das Chiffrat encrypted.bin (Binärdatei) ist mit RC4 verschlüsselt worden, der Schlüssel ist unbekannt. Ihre Aufgabe ist es, den Klartext wiederzugewinnen. Gehen Sie dafür wie folgt vor:
(a) Schreiben Sie eine Software für die Ver- und Entschlüsselung mit dem Stromchiffre RC4.
(b) Nutzen Sie Ihre RC4-Software für einen Brute-Force-Angriff auf den Chiffrattext. Starten Sie hierfür mit den Annahmen, dass
(i) der Schlüsselraum klein ist und (ii) alle verwendeten Schlüsselbytes Kleinbuchstaben (ASCII-Code) sind. Gehen Sie auch diesmal wieder davon aus, dass der Klartext ein Text einer natürlichen Sprache ist.

Implementierung, die man Copy Paste machen kann
https://github.com/manojpandey/rc4/blob/master/rc4-3.py

In [4]:
## Read in files
# Read in chiffrat
en = open('encryptedbin.sec', 'rb')
encrypt = en.read()


RC4 Verschlüsselung Implementierung

In [5]:
# Will use codecs, as 'str' object in Python 3 doesn't have any attribute 'decode'
import codecs


def initalize(key):
    #initalize is for initializing the key 

    #convert a single unicode character into its integer represntation
    key = [ord(c) for c in key]
    keylength = len(key)

    #make a list 
    S = list(range(256))
    
    
    j = 0
    for i in range(256):
        t = key[i % keylength]
        j = (j + S[i] + t) % 256
        #swap values
        S[i], S[j] = S[j], S[i]
    return S

def get_keystream(key):
    #get the keystream 

    S = initalize(key)
    i = 0
    j = 0
    while True:
        i = (i + 1) % 256
        j = (j + S[i]) % 256

        S[i], S[j] = S[j], S[i]  # swap values
        K = S[(S[i] + S[j]) % 256]
        yield K

def crypt_logic(text, key):
    #key needs to be in keystream already

    #make array for result
    result = []

    #for the entire text xor with keystream
    for counter in text:
       #xor the two strings
       #xored = chr(counter ^ (next(key)))
       xored = ("%02X" % (counter ^ (next(key))))
       result.append(xored)
    result = ''.join(result)
    #result is in hexform
    return result



def natural_language(result):
    #try making it an ascii string
    try:
        n_result = codecs.decode(result, 'hex_codec').decode('ASCII')
        #bei ä geht das nicht, weil kein ASCII Zeichen
        print(n_result)
        return True
    except:
        #if not ascii does not work out, reset xored
        #return 'try again'
        pass

def RC4_decrypt(ciphertext, key):
    #process of decrypting with RC4

    keystream = get_keystream(key)

    #make ascii ciphertext into hexcoded ciphertext
    #ciphertext = codecs.decode(ciphertext, 'hex_codec')
    
    res = crypt_logic(ciphertext, keystream)

    #try making it into natural language
    result = natural_language(res)
    return result


def RC4(ciphertext, key):
    #process of encrypting with RC4

    #convert a single unicode character into its integer represntation
    ciphertext = [ord(c) for c in ciphertext]

    #initialize and get keystream
    keystream = get_keystream(key)
    encrypt_text = crypt_logic(ciphertext, keystream)
    return encrypt_text


#test_text = 'TEST'
#test_key = 'Key'
#test_text ='BBF316E8D940AF0AD3'
#print(RC4_decrypt(test_text, test_key))


In [6]:
import itertools
from itertools import product
import string

#generate key from a... to z...
def generate_key_num(length):
    for combination in product(string.ascii_lowercase, repeat=length):
        yield list(combination)

def get_keys(length):
    keys = []
    key_generator = generate_key_num(length)
    
    #test
    #for i in key_generator:
      #  print (i)
    
    while True:
        try:
            keys.append(next(key_generator))
        except:
            break
    return keys

i = 2
for i in range(1, 6):
    print(i)
    canidates = get_keys(i)

    for j in canidates:
        #make array of characters into a string to use as a key
        key = ''.join(str(c) for c in j)
        if(RC4_decrypt(encrypt, key)):
            break
        
    


1
2
3
4
5
The DECT Standard Cipher (DSC) is a proprietary stream cipher used for encryption in the Digital Enhanced Cordless Telecommunications (DECT), which is a standard for short range cordless communication and widely deployed worldwide both in residential and enterprise environments. New weaknesses of the DSC stream cipher which are not discovered in previous works are explored and analyzed in this paper. Based on these weaknesses, new practical key recovery attacks and distinguishing attack on DSC with lower time cost are proposed. The first cryptanalytic result show that DSC can be broken in about 13.12 seconds in the known IV setting, when an offline phase that takes about 58.33 minutes is completed. After then, a distinguishing attack on DSC in the related key chosen IV setting is given, which has a time complexity of only 2 encryptions and a success probability of almost 1. Finally, based on the slide property, a key recovery attack on DSC with practical complexities is propo

1

2

3

4

5

The DECT Standard Cipher (DSC) is a proprietary stream cipher used for encryption in the Digital Enhanced Cordless Telecommunications (DECT), which is a standard for short range cordless communication and widely deployed worldwide both in residential and enterprise environments. New weaknesses of the DSC stream cipher which are not discovered in previous works are explored and analyzed in this paper. Based on these weaknesses, new practical key recovery attacks and distinguishing attack on DSC with lower time cost are proposed. The first cryptanalytic result show that DSC can be broken in about 13.12 seconds in the known IV setting, when an offline phase that takes about 58.33 minutes is completed. After then, a distinguishing attack on DSC in the related key chosen IV setting is given, which has a time complexity of only 2 encryptions and a success probability of almost 1. Finally, based on the slide property, a key recovery attack on DSC with practical complexities is proposed. The experimental result shows that DSC can be broken on a common PC within about 44.97 seconds in the multiple related key setting. The attacks on DSC proposed in this paper clearly show that a well-designed initialization is absolutely necessary to design a secure stream cipher.